### DISCUSSION TITLE
Corfu autocompletion popups are not working with eglot -- Presumably trying to turn off Eglot autocompletion completely

### DISCUSSION THREAD

**Discussion by swarnendubiswas** [joaotavora/eglot#1242] [Q&A]
I am unable to get autocompletion with `corfu` when I enable `eglot-grammarly`. Here is a MWE.

```
(setq straight-build-dir (format "build/%d%s%d"
                                 emacs-major-version
                                 version-separator
                                 emacs-minor-version)
      straight-use-package-by-default t)

(let ((bootstrap-file
       (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
      (bootstrap-version 6))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
        (url-retrieve-synchronously
         "https://raw.githubusercontent.com/raxod502/straight.el/develop/install.el"
         'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))

(setq use-package-hook-name-suffix nil)
(straight-use-package 'use-package)

(use-package corfu
  :custom
  (corfu-auto t "Enable auto completion")
  :config
  (global-corfu-mode 1))

(use-package cape
  :after corfu
  :config
  (add-hook
    'text-mode-hook
    (lambda ()
      (remove-hook 'completion-at-point-functions #'eglot-completion-at-point t)
      (setq-local completion-at-point-functions
        (list #'cape-file #'cape-dabbrev #'cape-dict)))))

(use-package eglot-grammarly
  :straight (:host github :repo "emacs-grammarly/eglot-grammarly")
  :hook
  (text-mode-hook . (lambda ()
      (require 'eglot-grammarly)
      (eglot-ensure))))
```

The value of `completion-at-point-functions` for a text file is as follows.

```
Its value is
(eglot-completion-at-point cape-file cape-dabbrev cape-dict)
Local in buffer email.txt; global value is
(tags-completion-at-point-function)
```

Please suggest what is wrong in the setup. Thanks.

**Comment by manuel-uberti** [comment]
Before making any suggestion (not because I don't want, but because I don't use either Straight, Corfu or eglot-grammarly so I wouldn't know where to start), can you try looking at https://joaotavora.github.io/eglot/#Troubleshooting-Eglot? Just so you can possibly narrow down the problem even further and make it easier to debug.

**Reply by joaotavora** [reply]
Especially because you didn't say what your want to get working and what exactly doesn't work.  Although this line is highly suspicious:
```
 (remove-hook 'completion-at-point-functions #'eglot-completion-at-point t)
```

Did ChatGPT give you this snippet :-) ? Don't trust it! Also, Corfu is known to be problematic with Eglot, I don't recommend it.

You can also try asking in Corfu.  They have a wiki there with an Eglot section.

**Reply by swarnendubiswas** [reply]
My guess is autocomplete popups with Corfu are not working because of wrong values for `completion-at-point-functions`. My goal is to set it to `(cape-file cape-dabbrev cape-dict)`.

I believe `eglot` is modifying `completion-at-point-functions` because of the LSP I have launched. Is there a way to prevent `eglot` from modifying `completion-at-point-functions` for language servers which are not supposed to provide completions but help with other tasks?

**Reply by manuel-uberti** [reply]
Off the top of my head can you try something like the following?
```emacs-lisp
(defun remove-eglot-completion-at-point ()
  (remove-hook 'completion-at-point-functions #'eglot-completion-at-point t))

(defun disable-eglot-completion ()
  (add-hook 'eglot-managed-mode-hook #'remove-eglot-completion-at-point t))

(add-hook 'text-mode-hook #'disable-eglot-completion)
```
This is untested and probably can be written better, but hopefully it can give you a start.

Still, try following the link I provided above to add more details and help debugging the problem a bit better. I am not sure my code above is the *right* solution for the issue you have.

**Reply by joaotavora** [reply]
@swarnendubiswas  It would help if you stated what you want to do.  So you don't want to use Eglot completion at all?

If so use something similar to what @manuel-uberti suggested.
